### DISCUSSION TITLE
How 'random' is shuffle? + what exactly is instant mix?

### DISCUSSION THREAD

**Discussion by Emporea** [jellyfin/jellyfin#9748] [Q&A]
I was wondering how 'random' shuffle is.

I don't want my music app to manipulate a basic randomiser into something that 'feels' more random, like Apple or Spotify do, so that similar songs don't come after each other.

Also I was wondering if there is some documenation on how the Instant mix works. Which metadata does it take into consideration? Genres? Something else?

**Comment by StergiosBinopoulos** [comment]
Instant mix only takes into consideration the genre. I made a plugin that attempts to improve the Instant Mix functionality. [BetterMix](https://github.com/StergiosBinopoulos/jellyfin-plugin-bettermix)

**Reply by JPVenson** [reply]
why not just properly intigrating that instead of using some dirty RP hacks?

**Reply by StergiosBinopoulos** [reply]
I'm not sure how I can overwrite the Jellyfin API

**Reply by StergiosBinopoulos** [reply]
@JPVenson as you noted a reverse proxy is not the best way to do this. I've updated the plugin to use an Action Filter instead. I'm not sure if this is the right way to do it. It still feels a bit hacky to me, but at least it's plug and play now. Is there a better way to do it?

**Reply by nielsvanvelzen** [reply]
You could contribute an interface to the server that the plugin can then implement

**Reply by StergiosBinopoulos** [reply]
Thank you @nielsvanvelzen. I will consider doing just that. I will have to look more into how I can achieve this, since I'm pretty new to C# and .NET.

**Reply by JPVenson** [reply]
i guess you could also just try and decorate the [`IMusicManager`](https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Controller/Library/IMusicManager.cs) and override the instant mix method with your own implementation

**Reply by StergiosBinopoulos** [reply]
@JPVenson this seems like a good solution and it would also work with older Jellyfin versions. I will give it a try. Thank you!

**Comment by Emporea** [comment]
Considering my original question, can someone reference the relevant code parts or explain how the current default shuffle and instant mix features work?

**Reply by StergiosBinopoulos** [reply]
It uses the music manager implementation Emby.Server.Implementations/Library/MusicManager.cs. The function used is GetInstantMixFromGenres. It returns a random shuffle of songs with matching genres.

**Reply by Emporea** [reply]
Okay but what happens if you shuffle all albums?

**Reply by StergiosBinopoulos** [reply]
I'm not sure you can do that. Maybe you are using a music client like Finamp. In that case the instant mix on albums doesn't really use the server side implementation. It would shuffle randomly all the albums you've selected. So you would get a random shuffle of all your songs. If you could do it in jellyfin (web) you would also get a random shuffle of all your songs, since you would have all available genres selected (except if an album and its songs have mismatched genre labels, if you manually edited it).
